File: backend/app/schemas/review.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


def _rating_field():
    return Field(None, ge=1, le=5)


class PropertyReviewCreateRequest(BaseModel):
    property_id: UUID
    tenancy_record_id: UUID
    rating_plumbing: int | None = _rating_field()
    rating_electricity: int | None = _rating_field()
    rating_water: int | None = _rating_field()
    rating_it_cabling: int | None = _rating_field()
    rating_hvac: int | None = _rating_field()
    rating_amenity_stove: int | None = _rating_field()
    rating_amenity_washer: int | None = _rating_field()
    rating_amenity_fridge: int | None = _rating_field()
    rating_infra_water_tank: int | None = _rating_field()
    rating_infra_irrigation: int | None = _rating_field()
    rating_health_dust: int | None = _rating_field()
    rating_health_breathing: int | None = _rating_field()
    rating_health_sewage: int | None = _rating_field()
    review_text: str = Field(min_length=20, max_length=5000)
    public_excerpt: str | None = Field(None, max_length=300)


class PropertyReviewResponse(BaseModel):
    id: UUID
    property_id: UUID
    tenant_id: UUID
    rating_plumbing: int | None
    rating_electricity: int | None
    rating_water: int | None
    rating_it_cabling: int | None
    rating_hvac: int | None
    rating_amenity_stove: int | None
    rating_amenity_washer: int | None
    rating_amenity_fridge: int | None
    rating_infra_water_tank: int | None
    rating_infra_irrigation: int | None
    rating_health_dust: int | None
    rating_health_breathing: int | None
    rating_health_sewage: int | None
    overall_rating: float
    review_text: str
    public_excerpt: str | None
    status: str
    verification_status: str
    is_flagged: bool
    published_at: datetime | None
    created_at: datetime
    photos: list["PhotoResponse"] = []

    model_config = {"from_attributes": True}


class PropertyReviewSnippetResponse(BaseModel):
    """Free tier: excerpt + aggregated scores only, no detailed text."""
    id: UUID
    property_id: UUID
    overall_rating: float
    public_excerpt: str | None
    status: str
    verification_status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PropertyReviewSummaryResponse(BaseModel):
    """Free tier: just averages and count."""
    property_id: UUID
    review_count: int
    avg_overall: float
    avg_plumbing: float | None
    avg_electricity: float | None
    avg_water: float | None
    avg_it_cabling: float | None
    avg_hvac: float | None
    avg_amenity_stove: float | None
    avg_amenity_washer: float | None
    avg_amenity_fridge: float | None
    avg_infra_water_tank: float | None
    avg_infra_irrigation: float | None
    avg_health_dust: float | None
    avg_health_breathing: float | None
    avg_health_sewage: float | None


class LandlordReviewCreateRequest(BaseModel):
    landlord_id: UUID
    property_id: UUID
    tenancy_record_id: UUID
    rating_responsiveness: int | None = _rating_field()
    rating_demeanor: int | None = _rating_field()
    rating_repair_payments: int | None = _rating_field()
    rating_availability: int | None = _rating_field()
    rating_payment_flexibility: int | None = _rating_field()
    review_text: str = Field(min_length=20, max_length=5000)


class LandlordReviewResponse(BaseModel):
    id: UUID
    landlord_id: UUID
    tenant_id: UUID
    property_id: UUID
    rating_responsiveness: int | None
    rating_demeanor: int | None
    rating_repair_payments: int | None
    rating_availability: int | None
    rating_payment_flexibility: int | None
    overall_rating: float
    review_text: str
    status: str
    verification_status: str
    is_flagged: bool
    published_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PhotoResponse(BaseModel):
    id: UUID
    file_url: str
    file_name: str
    sort_order: int

    model_config = {"from_attributes": True}

File: frontend/src/types/review.ts
export interface PropertyReview {
  id: string
  property_id: string
  tenant_id: string
  rating_plumbing: number | null
  rating_electricity: number | null
  rating_water: number | null
  rating_it_cabling: number | null
  rating_hvac: number | null
  rating_amenity_stove: number | null
  rating_amenity_washer: number | null
  rating_amenity_fridge: number | null
  rating_infra_water_tank: number | null
  rating_infra_irrigation: number | null
  rating_health_dust: number | null
  rating_health_breathing: number | null
  rating_health_sewage: number | null
  overall_rating: number
  review_text: string
  public_excerpt: string | null
  status: string
  verification_status: string
  is_flagged: boolean
  published_at: string | null
  created_at: string
  photos: PhotoItem[]
}

/** Snippet returned for unauthenticated / no-unlock users */
export interface PropertyReviewSnippet {
  id: string
  property_id: string
  overall_rating: number
  public_excerpt: string | null
  status: string
  verification_status: string
  created_at: string
}

export interface LandlordReview {
  id: string
  landlord_id: string
  tenant_id: string
  property_id: string
  rating_responsiveness: number | null
  rating_demeanor: number | null
  rating_repair_payments: number | null
  rating_availability: number | null
  rating_payment_flexibility: number | null
  overall_rating: number
  review_text: string
  status: string
  verification_status: string
  is_flagged: boolean
  published_at: string | null
  created_at: string
}

export interface PhotoItem {
  id: string
  file_url: string
  file_name: string
  sort_order: number
}

export interface ReviewSummary {
  property_id: string
  review_count: number
  avg_overall: number
  avg_plumbing: number | null
  avg_electricity: number | null
  avg_water: number | null
  avg_it_cabling: number | null
  avg_hvac: number | null
  avg_amenity_stove: number | null
  avg_amenity_washer: number | null
  avg_amenity_fridge: number | null
  avg_infra_water_tank: number | null
  avg_infra_irrigation: number | null
  avg_health_dust: number | null
  avg_health_breathing: number | null
  avg_health_sewage: number | null
}
